### PR TITLE
COG store creation

### PR DESCRIPTION
**Pull request by JakobMiksch**
- depends on 
  - https://github.com/meggsimum/geoserver-node-client/pull/112
  - https://github.com/klips-project/klips-api/pull/43
  - https://github.com/klips-project/klips-sdi/pull/96
- makes it possible to create an image mosaic store from COGs
- worker `geoserver-create-imagemosaic-datastore` 
  - can now create a COG based image mosaic
  - refactoring
- worker `geoserver-publish-imagemosaic`
  - can now publish COGs to a image mosaic
  - refactoring
- worker `geotiff-optimizer`
  - fix bug: actually create COG format and not GeoTIFF
- worker `geotiff-validtor`
  - move base image to GDAL image and link to existing GDAL. Now we can read all codecs

### CODE REVIEW

File: src/geoserver-create-imagemosaic-datastore/create-mosaic-store.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+ * @param {string} geoserverDataDir The path of the GeoServer data directory
+ */
+export const createClassicMosaicStore = async (grc, pgConf, ws, covStore, prototypeGranule, geoserverDataDir) => {
+
+  const { host, port, schema, database, user, password } = pgConf;
+  if (!host || !port || !schema || !database || !user || !password) {
+    throw 'PostgreSQL credentials are not complete.'
+  }
+
+  log(`CoverageStore ${covStore}  does not exist. Try to create it ...`);
```
Comment (hblitza) [inline src/geoserver-create-imagemosaic-datastore/create-mosaic-store.js:24]:
```suggestion
  log(`CoverageStore ${covStore} does not exist. Try to create it ...`);
```

**Review by hblitza — APPROVED** [review]
Thanks, lgtm. Just some double code parts in `createClassicMosaicStore` and `createCogMosaicStore`. not critical at all but might be enhanced in future.

**Comment by JakobMiksch** [comment]
> Just some double code parts in `createClassicMosaicStore` and `createCogMosaicStore`. not critical at all but might be enhanced in future.

True, I thought about this as well. Currently there are only two occurrences. If they become more I agree to refactor. Currently is borderline i would say.

File: src/geoserver-publish-imagemosaic/publish-granule.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    throw 'Granule with this timestamp already exists.';
+  }
+
+  await grc.imagemosaics.addGranuleByRemoteFile(
+    ws, covStore, coverageToAdd
+  );
+
+  const granuleRecognisedByGeoServer = await grc.imagemosaics.doesGranuleExist(ws, covStore, covStore, coverageToAdd);
+
+
```
Comment (JakobMiksch) [inline src/geoserver-publish-imagemosaic/publish-granule.js:27]:
```suggestion

```

File: src/geotiff-optimizer/index.js [hunk 3]
```diff
@@ -23,6 +23,12 @@ const callbackWorker = async (workerJob, inputs) => {
     const inputPath = inputs[0];
     const outputPath = inputs[1];
 
+    // TODO: check if GeoTIFF is already optimized
```
Comment (JakobMiksch) [inline src/geotiff-optimizer/index.js:26]:
```suggestion
```

File: src/geotiff-optimizer/index.js [hunk 4]
```diff
@@ -23,6 +23,11 @@ const callbackWorker = async (workerJob, inputs) => {
     const inputPath = inputs[0];
     const outputPath = inputs[1];
 
+
```
Comment (JakobMiksch) [inline src/geotiff-optimizer/index.js:26]:
```suggestion
```

File: src/geoserver-create-imagemosaic-datastore/create-mosaic-store.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  zip.addFile('timeregex.properties', Buffer.from(cogConfigFiles.timeregex));
+  const zipPath = '/tmp/init_cog.zip';
+  zip.writeZip(zipPath);
+
+  const autoconfigure = false;
+  if(debug) console.log('create mosaic store');
+  await grc.datastores.createImageMosaicStore(ws, covStore, zipPath, autoconfigure);
+
+  if(debug) console.log('add granule');
+  await grc.imagemosaics.addGranuleByRemoteFile(ws, covStore, prototypeGranule);
```
Comment (JakobMiksch) [inline src/geoserver-create-imagemosaic-datastore/create-mosaic-store.js:49]: prevent checking if file is uploaded

**Review by weskamm — APPROVED** [review]

**Comment by weskamm** [comment]
When calling this via workflow `create-cog-datastore.json` the worker errors with the following output:

image

same goes for the `add-single-cog` workflow. Do i need to create something before?

**Pull request merged**